### DISCUSSION TITLE
Trouble with ParsedMaterial

### DISCUSSION THREAD

**Discussion by AlexSacMorane** [idaholab/moose#25990] [Q&A Modules: General] [answered]
Dear all,

I am using MOOSE to solve a Phase-Field problem but I have trouble with a ParsedMaterial.
You will find my code in the following. I got an error from the `if` condition in the material `expected '=' after parameter name '0.03', got ','`

I would like to define a variable `kappa_psi` depending on the value of the variable `psi`. 
How can I correct my code?
I stay available
Thanks in advance
Regards
Alexandre 
```
---------------------------------------------------

[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 250
  ny = 250
  nz = 0
  xmin = -1.5
  xmax = 1.5
  ymin = -1.5
  ymax = 1.5
  elem_type = QUAD4
[]

[GlobalParams]
  outputs = exodus
[]

[Variables]
  active = 'psi'
  [./psi]
    order = FIRST
    family = LAGRANGE
    [./InitialCondition]
      type = SmoothCircleIC
      invalue = 1
      outvalue = 0
      radius = 1
      x1 = 0
      y1 = 0
      int_width = 0.15
    [../]
  [../]
[]

[Kernels]
  active = 'dpsidt ACBulk_psi ACInterface_psi'

  [./dpsidt]
    type = TimeDerivative
    variable = psi
  [../]
  [./ACBulk_psi]
    type = AllenCahn
    variable = psi
    mob_name = L_psi
    f_name = g_psi
  [../]
  [./ACInterface_psi]
    type = ACInterface
    variable = psi
    mob_name = L_psi
    kappa_name = kappa_psi
  [../]

[]

[Materials]
  active = 'consts var free_energy_psi'

  [./consts]
    type = GenericConstantMaterial
    prop_names  = 'L_psi'
    prop_values = '1'
  [../]

  [./free_energy_psi]
    type = DerivativeParsedMaterial
    block = 0
    f_name = g_psi
    args = 'psi'
    constant_names = 'A_psi B_psi C_psi'
    constant_expressions = '46.5 -83. 31.5'
    function = 'A_psi * psi^4+B_psi * psi^3+C_psi * psi^2'
    enable_jit = true
    derivative_order = 2
  [../]

  [./var]
    type = ParsedMaterial
    property_name = kappa_psi
    coupled_variables = 'psi'
    function = if(psi>0.5, 0.03, 0.01)
  [../]
[]


[Preconditioning]
  [./SMP]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  scheme = 'bdf2'
  solve_type = 'NEWTON'

  l_max_its = 20
  l_tol = 1.0e-4
  l_abs_tol = 1.0e-6

  nl_max_its = 10
  nl_rel_tol = 1.0e-4
  nl_abs_tol = 1.0e-6

  start_time = 0.0
  end_time   = 1

  [./TimeStepper]
    type = SolutionTimeAdaptiveDT
    dt = 0.1
  [../]
[]

[Outputs]
  execute_on = 'initial timestep_end'
  exodus = true
[]
```

**Comment by lindsayad** [comment, marked as answer]
Can you try putting quotes around your function?

**Reply by AlexSacMorane** [reply]
I tried with quotes `function = 'if(psi>0.5, 0.03, 0.01)'` but I get an error `Invalid function if(psi>0.5, 0.03, 0.01) in ParsedMaterialHelper. Syntax error: Unknown identifier`

**Reply by lindsayad** [reply]
You need to fix this line
```
function = 'A_psipsi^4+B_psipsi^3+C_psi*psi^2'
```
presumably you want a `*` between the `psi`s in `psipsi`?

**Reply by lindsayad** [reply]
In the future please paste the whole error as that will make it easier to help you

**Reply by AlexSacMorane** [reply]
Thanks for your answer. Unfortunately, the error is not coming from here (the `*` were here, github erased them editing my text).
You will find the whole error in the following.

-------------------------

*** ERROR ***
Invalid function
if(psi>0.5, 0.03, 0.01)

in ParsedMaterialHelper.
Syntax error: Unknown identifier

Stack frames: 19
0: libMesh::print_trace(std::ostream&)
1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
2: void mooseError<char const (&) [18], std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, char const (&) [27], char const*>(char const (&) [18], std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char&&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, char const (&) [27], char const*&&)
3: ParsedMaterialHelper<false>::functionParse(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<PostprocessorName, std::allocator<PostprocessorName> > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::vector<double, std::allocator<double> > const&)
4: ParsedMaterialTempl<false>::ParsedMaterialTempl(InputParameters const&)
5: /home/asacmora/projects/moose/framework/libmoose-opt.so.0(+0x15cc598) [0x7f8399bb3598]
6: Factory::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters const&, unsigned int, bool)
7: /home/asacmora/projects/moose/framework/libmoose-opt.so.0(+0x120bfeb) [0x7f83997f2feb]
8: FEProblemBase::addMaterialHelper(std::vector<MaterialWarehouse*, std::allocator<MaterialWarehouse*> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters&)
9: FEProblemBase::addMaterial(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters&)
10: Action::timedAct()
11: ActionWarehouse::executeActionsWithAction(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
12: ActionWarehouse::executeAllActions()
13: MooseApp::runInputFile()
14: MooseApp::run()
15: main
16: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f8391907d90]
17: __libc_start_main
18: /home/asacmora/projects/moose/modules/phase_field/phase_field-opt(+0x240f) [0x564355e1840f]
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor

------------------------------------------

It seems Moose does not recognize `psi`, even if it is declared in the coupled variables.

**Reply by lindsayad** [reply]
I ran your input yesterday. When I put quotes around 'if(psi>0.5, 0.03, 0.01)' and made the suggested edit to the other parsed function, it ran fine for me

**Reply by AlexSacMorane** [reply]
Thanks for your time and your discussion.
